src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs

```diff
@@ -440,11 +440,11 @@ private static unsafe bool ParseNumber(ref char* str, char* strEnd, NumberStyles
                 int digEnd = 0;
                 while (true)
                 {
-                    if (char.IsAsciiDigit(ch) || (((options & NumberStyles.AllowHexSpecifier) != 0) && char.IsBetween((char)(ch | 0x20), 'a', 'f')))
+                    if ((options & NumberStyles.AllowBinarySpecifier) == 0 ? char.IsAsciiDigit(ch) || ((options & NumberStyles.AllowHexSpecifier) != 0 && char.IsBetween((char)(ch | 0x20), 'a', 'f')) : char.IsBetween(ch, '0', '1'))
                     {
                         state |= StateDigits;
 
-                        if (ch != '0' || (state & StateNonZero) != 0 || (bigNumber && ((options & NumberStyles.AllowHexSpecifier) != 0)))
+                        if (ch != '0' || (state & StateNonZero) != 0 || (bigNumber && (options & (NumberStyles.AllowHexSpecifier | NumberStyles.AllowBinarySpecifier)) != 0))
```

Comment:
love all the new tests :)

Reply:
> Should we hoist this `options` test

Got your point. As you found all the existing options tests code were inside whole loop, and during run time some option tests may happen only when data is met, so not sure whether we should hoist 'options' test which will turn to be called once but always happen. That is why I did not change this convention :)

> I am also not sure what's going on with the `bigNumber &&` here.

I am also not sure, so that I just added Binary stuff behind it like Hex :)

Reply:
Splitting it apart for readability is generally helpful. The logic is getting quite verbose to have on a single line IMO.

For the primitive types, we have it split apart into some early checks that go to `TryParseBinaryInteger*Style`: https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs,437

The check for "IsValidChar" can then be specialized by use of the generics and made a lot cheaper.

Reply:
@tannergooding Fixed. After introducing Hex/Bin parser types, here simplifies not only [prev line 447](https://github.com/dotnet/runtime/blob/6cea91abae2b9e7675e9f9b652ad2a0b1111ba9b/src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs#L447), but also [prev line 443](https://github.com/dotnet/runtime/blob/6cea91abae2b9e7675e9f9b652ad2a0b1111ba9b/src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs#L443) which involved hex/binary checks previously. Please check, thanks.